### ISSUE TITLE
cargo test fails on nightly

### ISSUE THREAD

**Issue by stanciuadrian** [rust-analyzer/rust-analyzer#8585]
Windows log

```shell
> rustup show
Default host: x86_64-pc-windows-msvc
rustup home:  C:\Users\X\.rustup

installed toolchains
--------------------

stable-x86_64-pc-windows-msvc
nightly-x86_64-pc-windows-msvc (default)

active toolchain
----------------

nightly-x86_64-pc-windows-msvc (default)
rustc 1.53.0-nightly (5a4ab2645 2021-04-18)

> cargo test
[...]

test tests::stmt_parser_tests ... ok
test tests::parser_tests ... FAILED
test tests::parser_fuzz_tests ... ok
test tests::lexer_tests ... FAILED
test tests::self_hosting_parsing ... ok

failures:

---- tests::parser_tests stdout ----

error: expect test failed
   --> C:\ra\rust-analyzer\crates/syntax/test_data\parser/inline/ok\0003_where_pred_for.rast

You can update all `expect![[]]` tests by running:

    env UPDATE_EXPECT=1 cargo test

To update a single test, place the cursor on `expect` token and use `run` feature of rust-analyzer.

Expect:
----
              LIFETIME_IDENT@31..33 "\'a"
                          LIFETIME_IDENT@42..44 "\'a"
----

Actual:
----
              LIFETIME_IDENT@31..33 "'a"
                          LIFETIME_IDENT@42..44 "'a"
----

---- tests::lexer_tests stdout ----


error: expect test failed
   --> C:\ra\rust-analyzer\crates/syntax/test_data\lexer/ok\0006_chars.txt

Expect:
----
CHAR 3 "\'x\'"
WHITESPACE 1 " "
CHAR 3 "\' \'"
WHITESPACE 1 " "
CHAR 3 "\'0\'"
WHITESPACE 1 " "
CHAR 7 "\'hello\'"
WHITESPACE 1 " "
CHAR 6 "\'\\x7f\'"
WHITESPACE 1 " "
CHAR 4 "\'\\n\'"
WHITESPACE 1 " "
CHAR 4 "\'\\\\\'"
WHITESPACE 1 " "
CHAR 4 "\'\\\'\'"
WHITESPACE 1 "\n"

----

Actual:
----
CHAR 3 "'x'"
WHITESPACE 1 " "
CHAR 3 "' '"
WHITESPACE 1 " "
CHAR 3 "'0'"
WHITESPACE 1 " "
CHAR 7 "'hello'"
WHITESPACE 1 " "
CHAR 6 "'\\x7f'"
WHITESPACE 1 " "
CHAR 4 "'\\n'"
WHITESPACE 1 " "
CHAR 4 "'\\\\'"
WHITESPACE 1 " "
CHAR 4 "'\\''"
WHITESPACE 1 "\n"

----

failures:
    tests::lexer_tests
    tests::parser_tests

test result: FAILED. 49 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.00s

error: test failed, to rerun pass '-p syntax --lib'
```

WSL 2 log:

```
$ rustup show
Default host: x86_64-unknown-linux-gnu
rustup home:  /home/X/.rustup

installed toolchains
--------------------

stable-x86_64-unknown-linux-gnu
nightly-x86_64-unknown-linux-gnu (default)

active toolchain
----------------

nightly-x86_64-unknown-linux-gnu (default)
rustc 1.53.0-nightly (5a4ab2645 2021-04-18)

$ cargo test
test diagnostics::to_proto::tests::handles_macro_location ... FAILED
test semantic_tokens::tests::test_diff_insert_at_beginning ... ok
test semantic_tokens::tests::test_diff_insert_in_middle ... ok
test semantic_tokens::tests::test_diff_remove_from_beginning ... ok
test semantic_tokens::tests::test_diff_remove_from_end ... ok
test semantic_tokens::tests::test_diff_remove_from_middle ... ok
test diagnostics::to_proto::tests::rustc_incompatible_type_for_trait ... FAILED

Expect:
            message: "can\'t compare `{integer}` with `&str`\nthe trait `std::cmp::PartialEq<&str>` is not implemented for `{integer}
Actual:
            message: "can't compare `{integer}` with `&str`\nthe trait `std::cmp::PartialEq<&str>` is not implemented for `{integer}`
Expect:
            message: "method `next` has an incompatible type for trait\nexpected type `fn(&mut ty::list_iter::ListIterator<\'list, M>) -> std::option::Option<&ty::Ref<M>>`\n   found type `fn(&ty::list_iter::ListIterator<\'list, M>) -> std::option::Option<&\'list ty::Ref<M>>`",
Actual:
            message: "method `next` has an incompatible type for trait\nexpected type `fn(&mut ty::list_iter::ListIterator<'list, M>) -> std::option::Option<&ty::Ref<M>>`\n   found type `fn(&ty::list_iter::ListIterator<'list, M>) -> std::option::Option<&'list ty::Ref<M>>`",
```

I understand there's a change in rustc `nightly` that's not in `stable` yet. Switching to `stable` passes all the tests.

**Comment by stanciuadrian** [comment]
It seems that the expected string is not generated in a portable/stable way.
The same code-base has different failing tests on different OS-es.

**Comment by lnicola** [comment]
This is caused by a bugfix in the `Debug` implementation for `str`. It used to (wrongly) escape `'`.

**Comment by stanciuadrian** [comment]
I can prepare a pull request for the next rustc `stable`.

**Comment by lnicola** [comment]
Closed by #9315.
